internal/app/singularity/registry.go

```diff
@@ -0,0 +1,146 @@
+package singularity
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+
+	scs "github.com/sylabs/scs-library-client/client"
+	"github.com/sylabs/singularity/internal/pkg/client/cache"
+	"github.com/sylabs/singularity/internal/pkg/library"
+	"github.com/sylabs/singularity/internal/pkg/sylog"
+	"github.com/sylabs/singularity/internal/pkg/util/uri"
+	"github.com/sylabs/singularity/pkg/signing"
+)
+
+var errNotInCache = fmt.Errorf("image was not found in cache")
+
+// Registry represents image registries and common operations
+// performed, e.g pull and push.
+type Registry interface {
+	Pull(ctx context.Context, from, to string) error
+}
+
+// Library is a Registry implementation for Sylabs Cloud Library.
+type Library struct {
+	keystoreURI string
+
+	client *scs.Client
+	cache  *cache.Handle
+}
+
+// NewLibrary initializes and returns new Library ready to  be used.
+func NewLibrary(client *scs.Client, cache *cache.Handle, keystoreURI string) *Library {
+	return &Library{
+		keystoreURI: keystoreURI,
+		client:      client,
```

Comment:
That will require creation some kind of wrapper for `*scs.Client` b/c `DownloadImage` is a function in Singularity repo, which accepts `*scs.Client`. 
Also code readability will suffer, so I'd prefer to keep e2e tests and bare `*scs.Client` rather than introducing this new interface.

Reply:
I probably did not explain myself correctly.

It's enough to change the code like this:

```
diff --git a/internal/pkg/library/library.go b/internal/pkg/library/library.go
index 5ecde1b65..fa03d2fa8 100644
--- a/internal/pkg/library/library.go
+++ b/internal/pkg/library/library.go
@@ -29,8 +29,12 @@ func NormalizeLibraryRef(libraryRef string) string {
        return ir
 }

+type ImageDownloader interface {
+       DownloadImage(ctx context.Context, w io.Writer, path, tag string, callback func(int64, io.Reader, io.Writer) error) error
+}
+
 // DownloadImage is a helper function to wrap library image download operation
-func DownloadImage(ctx context.Context, c *client.Client, imagePath, libraryRef string, callback progressCallback) error {
+func DownloadImage(ctx context.Context, c ImageDownloader, imagePath, libraryRef string, callback progressCallback) error {
        // reassemble "stripped" library ref for scs-library-client
        validLibraryRef := "library:///" + libraryRef

@@ -66,12 +70,16 @@ func DownloadImage(ctx context.Context, c *client.Client, imagePath, libraryRef

 // DownloadImageNoProgress downloads an image from the library without
 // displaying a progress bar while doing so
-func DownloadImageNoProgress(ctx context.Context, c *client.Client, imagePath, libraryRef string) error {
+func DownloadImageNoProgress(ctx context.Context, c ImageDownloader, imagePath, libraryRef string) error {
        return DownloadImage(ctx, c, imagePath, libraryRef, nil)
 }

+type ImageSearcher interface {
+       Search(ctx context.Context, args map[string]string) (*client.SearchResults, error)
+}
+
 // SearchLibrary searches the library and outputs results to stdout
-func SearchLibrary(ctx context.Context, c *client.Client, value string) error {
+func SearchLibrary(ctx context.Context, c ImageSearcher, value string) error {
        if len(value) < 3 {
                return fmt.Errorf("bad query '%s'. You must search for at least 3 characters", value)
        }
```

you don't need other changes (and just for the record, you don't need to export the interfaces either). The rest of the code doesn't need to care about this, it can continue using a client.

(obviously this refers to the existing code, not this PR)

The point of this change is that you can write a test for these functions without requiring an actual client that will hit the network. We already have enough problems with "unit" tests hitting the network and this is a simple enough change that would help with that.

Reply:
And yes, I provided the example in the wrong package, sorry :-\

Reply:
Looks like a separate issue/PR, do you mind creating one?